### PR TITLE
Fix for Support selection of datastore for dynamic provisioning in vSphere

### LINKED ISSUE #40558: Support selection of datastore for dynamic provisioning in vSphere
Current vSphere Cloud provider doesn't allow a user to select a datastore for dynamic provisioning. All the volumes are created in default datastore provided by the user in the global vsphere configuration file. 

With this fix, the user will be able to provide the datastore in the storage class definition. This will allow the volumes to be created in the datastore specified by the user in the storage class definition. This field is optional. If no datastore is specified, the volume will be created in the default datastore specified in the global config file.

For example:

User creates a storage class with the datastore
---
> kind: StorageClass
> apiVersion: storage.k8s.io/v1beta1
> metadata:
>   name: slow
> provisioner: kubernetes.io/vsphere-volume
> parameters:
>     diskformat: thin
>     datastore: VMFSDatastore

Now the volume will be created in the datastore - "VMFSDatastore" specified by the user. 

If the user creates a storage class without any datastore
---
> kind: StorageClass
> apiVersion: storage.k8s.io/v1beta1
> metadata:
>   name: slow
> provisioner: kubernetes.io/vsphere-volume
> parameters:
>     diskformat: thin

Now the volume will be created in the datastore which in the global configuration file (vsphere.conf)

### PR DESCRIPTION
**Pull request by BaluDontu**
Fixes #40558

Current vSphere Cloud provider doesn't allow a user to select a datastore for dynamic provisioning. All the volumes are created in default datastore provided by the user in the global vsphere configuration file.

With this fix, the user will be able to provide the datastore in the storage class definition. This will allow the volumes to be created in the datastore specified by the user in the storage class definition. This field is optional. If no datastore is specified, the volume will be created in the default datastore specified in the global config file.

For example:

User creates a storage class with the datastore

kind: StorageClass
apiVersion: storage.k8s.io/v1beta1
metadata:
name: slow
provisioner: kubernetes.io/vsphere-volume
parameters:
diskformat: thin
datastore: VMFSDatastore
Now the volume will be created in the datastore - "VMFSDatastore" specified by the user.

If the user creates a storage class without any datastore

kind: StorageClass
apiVersion: storage.k8s.io/v1beta1
metadata:
name: slow
provisioner: kubernetes.io/vsphere-volume
parameters:
diskformat: thin
Now the volume will be created in the datastore which in the global configuration file (vsphere.conf)

@pdhamdhere @kerneltime

### CODE REVIEW

**Comment by wojtek-t** [comment]
I don't have enough context to review it.

Maybe @luomiao @kerneltime or @imkin ?

File: pkg/cloudprovider/providers/vsphere/vsphere.go [hunk 1]
```diff
@@ -631,36 +633,36 @@ func (vs *VSphere) ScrubDNS(nameservers, searches []string) (nsOut, srchOut []st
...
 	// Fetch and set data center
 	dc, err := f.Datacenter(ctx, cfg.Global.Datacenter)
 	if err != nil {
-		return nil, nil, nil, nil, err
+		return nil, nil, nil, err
 	}
 	f.SetDatacenter(dc)
 
 	// Find datastores
 	ds, err := f.Datastore(ctx, cfg.Global.Datastore)
```
Comment (kerneltime) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: This looks like dead code

File: pkg/cloudprovider/providers/vsphere/vsphere.go [hunk 2]
```diff
@@ -680,6 +682,14 @@ func cleanUpController(ctx context.Context, newSCSIController types.BaseVirtualD
 	return nil
 }
 
+// Get datastore from VMDisk path.
+// vmDiskPath format is [<datastore> kubevols/<volume>]
+func getDatastoreFromVMDiskPath(vmDiskPath string) string {
```
Comment (kerneltime) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: We should update the vendor sw and pick up this change https://github.com/vmware/govmomi/commit/e152c3553e1101445babdb94082a36df9fd3ccf3

File: pkg/cloudprovider/providers/vsphere/vsphere.go [hunk 3]
```diff
@@ -1190,6 +1212,14 @@ func (vs *VSphere) DetachDisk(volPath string, nodeName k8stypes.NodeName) error
 func (vs *VSphere) CreateVolume(volumeOptions *VolumeOptions) (volumePath string, err error) {
 
 	var diskFormat string
+	var datastore string
+
+	// Default diskformat as 'thin'
+	if volumeOptions.Datastore == "" {
+		datastore = vs.cfg.Global.Datastore
```
Comment (kerneltime) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: We should update samples posted in this repo to reflect this change.

**Review by kerneltime — COMMENTED** [review]
Overall looks good but a few changes are needed. We should take this opportunity to update the vendor govmomi code as well.

File: pkg/cloudprovider/providers/vsphere/vsphere.go (hunk 2 above)
Comment (kerneltime) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: I don't want to replicate the string parsing logic given that govmomi has helpers for it now.

**Comment by kerneltime** [comment]
@wojtek-t thanks! I will continue review this

**Comment by kerneltime** [comment]
@k8s-bot ok to test

File: pkg/cloudprovider/providers/vsphere/vsphere.go [hunk 4]
```diff
@@ -1190,6 +1212,14 @@ func (vs *VSphere) DetachDisk(volPath string, nodeName k8stypes.NodeName) error
 func (vs *VSphere) CreateVolume(volumeOptions *VolumeOptions) (volumePath string, err error) {
 
 	var diskFormat string
+	var datastore string
+
+	// Default diskformat as 'thin'
```
Comment (divyenpatel) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: Update comment to mention about default datastore as vs.cfg.Global.Datastore

**Comment by thockin** [comment]
This is entriely withing vsphere, so I am happy to review for sanity but can't review for correctness

**Comment by thockin** [comment]
@k8s-bot ok to test

**Comment by kerneltime** [comment]
Thanks @thockin, @divyenpatel and I will take care of correctness.

**Pull request closed without merging**